cmake/AwsCheckHeaders.cmake

```diff
@@ -15,57 +15,84 @@ endif()
-            target_compile_options(${HEADER_CHECKER_LIB} PRIVATE -Wall -Wextra -Wpedantic -Werror)
-        endif()
+    # We want to be able to verify that the proper C++ header guards are in place, so
+    # build this target as a C++ application
+    set_target_properties(${HEADER_CHECKER_LIB} PROPERTIES
+        LINKER_LANGUAGE CXX
+        CXX_STANDARD ${std}
+        CXX_STANDARD_REQUIRED 0
+        C_STANDARD 99
+    )
 
-        foreach(header IN LISTS ARGN)
-            if (NOT ${header} MATCHES "\\.inl$")
-                # create unique token for this file, e.g.:
-                # "${CMAKE_CURRENT_SOURCE_DIR}/include/aws/common/byte_buf.h" -> "aws_common_byte_buf_h"
-                file(RELATIVE_PATH include_path "${CMAKE_CURRENT_SOURCE_DIR}/include" ${header})
-                # replace non-alphanumeric characters with underscores
-                string(REGEX REPLACE "[^a-zA-Z0-9]" "_" unique_token ${include_path})
-                set(c_file "${HEADER_CHECKER_ROOT}/headerchecker_${unique_token}.c")
-                set(cpp_file "${HEADER_CHECKER_ROOT}/headerchecker_${unique_token}.cpp")
-                # include header twice to check for include-guards
-                # define a unique int or compiler complains that there's nothing in the file
+    # Ensure our headers can be included by an application with its warnings set very high
+    if(MSVC)
+        target_compile_options(${HEADER_CHECKER_LIB} PRIVATE /W4 /WX)
+    else()
+        target_compile_options(${HEADER_CHECKER_LIB} PRIVATE -Wall -Wextra -Wpedantic -Werror)
+    endif()
+
+    foreach(header IN LISTS ARGN)
+        if (NOT ${header} MATCHES "\\.inl$")
+            # create unique token for this file, e.g.:
+            # "${CMAKE_CURRENT_SOURCE_DIR}/include/aws/common/byte_buf.h" -> "aws_common_byte_buf_h"
+            file(RELATIVE_PATH include_path "${CMAKE_CURRENT_SOURCE_DIR}/include" ${header})
+            # replace non-alphanumeric characters with underscores
+            string(REGEX REPLACE "[^a-zA-Z0-9]" "_" unique_token ${include_path})
+            set(c_file "${HEADER_CHECKER_ROOT}/headerchecker_${unique_token}.c")
+            set(cpp_file "${HEADER_CHECKER_ROOT}/headerchecker_${unique_token}.cpp")
+            # include header twice to check for include-guards
+            # define a unique int or compiler complains that there's nothing in the file
+            if(disable_c_header_checker)
+                file(WRITE "${cpp_file}" "#include <${include_path}>\n#include <${include_path}>\nint ${unique_token}_cpp;")
+                target_sources(${HEADER_CHECKER_LIB} PUBLIC "${cpp_file}")
+            else()
                 file(WRITE "${c_file}" "#include <${include_path}>\n#include <${include_path}>\nint ${unique_token}_c;\n")
                 file(WRITE "${cpp_file}" "#include <${include_path}>\n#include <${include_path}>\nint ${unique_token}_cpp;")
-
                 target_sources(${HEADER_CHECKER_LIB} PUBLIC "${c_file}" "${cpp_file}")
```

Comment:
you can call target_sources() multiple times, it's additive
```
file(WRITE "${cpp_file}" "#include <${include_path}>\n#include <${include_path}>\nint ${unique_token}_cpp;")
target_sources(${HEADER_CHECKER_LIB} PUBLIC "${cpp_file}")

# Don't need to check C++ headers from C
if (NOT is_cxx)
    file(WRITE "${c_file}" "#include <${include_path}>\n#include <${include_path}>\nint ${unique_token}_c;\n")
    target_sources(${HEADER_CHECKER_LIB} PUBLIC "${c_file}"
endif()
```

Reply:
Thanks, fixed.